### PR TITLE
Fix: CapRover Dashboard App: Details of Workers disappear after few seconds

### PR DESCRIPTION
**Pull request by Mahmoud-Emad**
### Description

Fix listing Caprover workers

### Changes

- Send both of the deployments, leader, and workers.
- Call the 'getGrafanaUrl' on mount the component.
- Delete the full cluster.
- Ensure that workers are correctly removed from the deployment list when the leader is deleted to maintain data integrity.
- Merged the latest changes from the development branch to keep the codebase up to date.
- Conducted thorough testing to verify the complete functionality and ensure no regressions were introduced.

### Test Scenarios

1. Deploy `leader`, and `worker`, Go go to the contracts page and delete the deployed `leader`.
2. Deploy `leader`, attach a `worker` to it then delete the whole cluster just in time without refreshing the page.
3. Deploy `leader`, and `worker`, delete the deployed `worker` and attach another, then delete the whole cluster just in time without refreshing the page.
4. Deploy a cluster with 1 leader and 2 workers check the deployment listing dialog and make sure all of them are listed
3. Delete a cluster with the leader and more than 3 workers and make sure all of them are deleted.

PS: Make sure no errors on the Caprover listing page.

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3281

### Screenshots/Videos

[Screencast from 09-03-2024 05:27:14 PM.webm](https://github.com/user-attachments/assets/45752f9d-d688-4869-8327-f68c5aad1c65)

image

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

### CODE REVIEW

File: packages/playground/src/weblets/tf_deployment_list.vue [hunk 1]
```diff
@@ -464,53 +474,114 @@ const deletingDialog = ref(false);
 const table = ref() as Ref<{ loadDeployments(): void }>;
 const gridStore = useGrid();
 const grid = gridStore.client as GridClient;
+const hasWorkers = computed(
+  () => selectedItems.value.map(item => item.projectName.includes(ProjectName.Caprover.toLocaleLowerCase())).length > 0,
+);
```
Comment (MohamedElmdary) [inline packages/playground/src/weblets/tf_deployment_list.vue]: I think you mean filter instead of map?

**Review by maayarosama — DISMISSED** [review]

**Review by 0oM4R — COMMENTED** [review]
I was able to list and delete all workers fine, but on adding new one i got this error
Screenshot from 2024-09-12 12-06-51

**Comment by Mahmoud-Emad** [comment]
> I was able to list and delete all workers fine, but on adding new one i got this error Screenshot from 2024-09-12 12-06-51

This happened because the worker is listed instead of the master, which is wrong

**Review by zaelgohary — COMMENTED** [review]
Now listing works fine as it lists the leader not the worker, however, on adding another worker after deploying, I faced this error.

image

**Comment by Mahmoud-Emad** [comment]
> Now listing works fine as it lists the leader not the worker, however, on adding another worker after deploying, I faced this error.
> 
> image

This issue isn't related to my PR, it could be another issue, it says there is another network workload with the same name already exists.

File: packages/playground/src/weblets/tf_deployment_list.vue [hunk 2]
```diff
@@ -476,53 +486,112 @@ const deletingDialog = ref(false);
...
-    table.value?.loadDeployments();
-  } catch (e) {
-    createCustomToast((e as Error).message, ToastType.danger);
-  } finally {
-    selectedItems.value = [];
-    deleting.value = false;
+  } catch (e: any) {
+    logDeletionError(item.name, e);
   }
 }
```
Comment (zaelgohary) [inline packages/playground/src/weblets/tf_deployment_list.vue]: Do we need to separate these into 3 functions?

**Comment by maayarosama** [comment]
Why is most of the code separated into different function and only used once, for example ```prepareForDeletion``` and ```finalizeDeletion```. Is it because this's cleaner code somehow??

**Comment by maayarosama** [comment]
I deployed a cluster and added two workers successfully. Listing of workers is fine
Screenshot from 2024-09-24 13-15-56
Screenshot from 2024-09-24 13-16-21

However on shoe details, workers are not shown
Screenshot from 2024-09-24 13-24-58

**Review by zaelgohary — CHANGES_REQUESTED** [review]
I couldn't view domain deployment on this branch but it worked normally on staging.

image

image

**Comment by Mahmoud-Emad** [comment]
@zaelgohary 


Working well with me, please make sure that you have the latest code and do  `yarn && make build`


https://github.com/user-attachments/assets/21131a58-1886-4db6-b3ec-89e1c52046e8

File: packages/playground/src/weblets/tf_deployment_list.vue (hunk 2 above)
Comment (Mahmoud-Emad) [reply packages/playground/src/weblets/tf_deployment_list.vue:542]: Why not?

File: packages/playground/src/utils/deploy_helpers.ts [hunk 3]
```diff
@@ -11,6 +11,30 @@ export function createNetwork(network: Network = {}): NetworkModel {
   return nw;
 }
 
+export function setCaproverWorkers(vms: ZmachineData[], projectName: string | undefined = undefined): ZmachineData {
```
Comment (zaelgohary) [inline packages/playground/src/utils/deploy_helpers.ts:14]: Why do we expect a projectName while we're setting caprover workers?

Comment (zaelgohary) [reply packages/playground/src/utils/deploy_helpers.ts:14]: Won't the project in this case always be Caprover?

**Comment by zaelgohary** [comment]
> Working well with me, please make sure that you have the latest code and do `yarn && make build`
> 
>  Screencast.from.2024-09-24.15-13-48.mp4

Can you click on show details as the error appears once clicking it?

I pulled latest changes and did yarn & make build.

image

File: packages/playground/src/utils/deploy_helpers.ts (hunk 3 above)
Comment (Mahmoud-Emad) [reply packages/playground/src/utils/deploy_helpers.ts:14]: > Why do we expect a projectName while we're setting caprover workers?

if i understand well, we save the workers as the leader name to be able to load them, so why don't I hard-code it? I don't want to set it in all of the calls. I want to set it only after deploying a new worker

**Comment by zaelgohary** [comment]
Workers are being showed as leaders in contracts page.

image

image



image

**Comment by zaelgohary** [comment]
I deleted the caprover deployment but 1 network contract still exist after deletion. Happened more than once.

image

**Comment by AhmedHanafy725** [comment]
> Workers are being showed as leaders in contracts page.
> 
> image
> 
> image
> 
> image

@zaelgohary In the contracts, all the cluster's workers carry the same name as the cluster (which is the leader name) so we can list them as part of the cluster

**Comment by Mahmoud-Emad** [comment]
> I deleted the caprover deployment but 1 network contract still exist after deletion. Happened more than once.
> 
> image

It's a known issue i think https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3425

**Comment by zaelgohary** [comment]
> > I deleted the caprover deployment but 1 network contract still exist after deletion. Happened more than once.
> > image
> 
> It's a known issue i think #3425

It's not. It was fixed in https://github.com/threefoldtech/tfgrid-sdk-ts/pull/3456 which was merged yesterday. Can you merge development to your branch? and if the issue still exist, we can open an issue for it as this only happen in Caprover.

**Comment by Mahmoud-Emad** [comment]
The issue is already solved yeah, I just noticed that I have to build the client first 

image

However, yeah one contract still exists and should be deleted!

https://github.com/threefoldtech/tfgrid-sdk-ts/issues/3473

**Review by zaelgohary — APPROVED** [review]

File: packages/playground/src/weblets/tf_deployment_list.vue [hunk 4]
```diff
@@ -484,29 +495,39 @@ watch(activeTab, () => (selectedItems.value = []));
...
+              deploymentName: item.deploymentName,
+              name: k8s ? item.deploymentName : item.name,
+              projectName: item.projectName,
+              ip: item.interfaces?.[0]?.ip,
+              k8s,
+            });
+          }
+        } catch (e: any) {
+          createCustomToast(`Failed to delete deployment with name: ${item.name}`, ToastType.danger);
+          console.log("Error while deleting deployment", e.message);
```
Comment (0oM4R) [inline packages/playground/src/weblets/tf_deployment_list.vue]: please use console.error to make it clear in the logger
```suggestion
          console.error("Error while deleting deployment", e.message);
```

**Review by zaelgohary — CHANGES_REQUESTED** [review]
I tried listing a deployment I have but got the following errors. Am I doing anything wrong?

image

image

**Comment by zaelgohary** [comment]
I deployed a new instance & whenever I view its details I get the following error:

image

I also got the same error when deleting a worker.

**Comment by maayarosama** [comment]
I tried adding workers most of the times it passed, but it failed once. Deletion works fine
Screenshot from 2024-09-29 13-22-25
Screenshot from 2024-09-29 13-24-07

**Review by zaelgohary — APPROVED** [review]
Works fine adding/deleting clusters & works.

image

**Review by maayarosama — APPROVED** [review]

**Pull request merged**